corehq/motech/repeaters/tasks.py

```diff
@@ -215,6 +226,198 @@ def _process_repeat_record(repeat_record):
         )
 
 
+@periodic_task(
+    run_every=PROCESS_REPEATERS_INTERVAL,
+    queue=settings.CELERY_PERIODIC_QUEUE,
+)
+def process_repeaters():
+    """
+    Processes repeaters, instead of processing repeat records
+    independently the way that ``check_repeaters()`` does.
+    """
+    process_repeater_lock = get_redis_lock(
+        PROCESS_REPEATERS_KEY,
+        timeout=24 * 60 * 60,
+        name=PROCESS_REPEATERS_KEY,
+    )
+    if not process_repeater_lock.acquire(blocking=False):
+        return
+    try:
+        for domain, repeater_id, lock_token in iter_ready_repeater_ids_forever():
+            process_repeater.delay(domain, repeater_id, lock_token)
+    finally:
+        process_repeater_lock.release()
+
+
+def iter_ready_repeater_ids_forever():
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for domain, repeater_id in iter_ready_repeater_ids_once():
+            if not toggles.PROCESS_REPEATERS.enabled(domain, toggles.NAMESPACE_DOMAIN):
```

Comment:
In a loop like this it's probably better to hold the full list of enabled domains in memory and check access that way. You can refresh the list every few seconds.

Reply:
:thinking: I've just checked, and it isn't cached, but somehow I thought it was. A list is a great idea.

Reply:
It's cached way down here: https://github.com/dimagi/commcare-hq/blob/master/corehq/toggles/models.py#L35

You can use `toggles.PROCESS_REPEATERS.get_enabled_domains()` to get the list of domains.

Reply:
db093d0eac0 This will refresh the list of domains every time it fetches repeater IDs by domain. It should be equivalent to filtering using the query, but avoid [very big queries](https://github.com/dimagi/commcare-hq/pull/34946#discussion_r1714375956) when a lot of domains are enabled.

Reply:
The incremental rollout might be very useful on production. https://github.com/dimagi/commcare-hq/pull/35033/commits/418ed3afe7f853edeeb1160bd025608ce85a0c99